File: apps/solana-programs/programs/market-settlement/src/events.rs
use anchor_lang::prelude::*;

/// Event emitted when proof is submitted
#[event]
pub struct ProofSubmitted {
    pub market: Pubkey,
    pub proof_id: u64,
    pub submitter: Pubkey,
    pub outcome_index: u8,
    pub proof_data: String,
    pub timestamp: i64,
}

/// Event emitted when proof is verified
#[event]
pub struct ProofVerified {
    pub market: Pubkey,
    pub proof_id: u64,
    pub verifier: Pubkey,
    pub is_valid: bool,
    pub timestamp: i64,
}

/// Event emitted when market is resolved
#[event]
pub struct MarketResolved {
    pub market: Pubkey,
    pub resolver: Pubkey,
    pub winning_outcome: u8,
    pub total_payout: u64,
    pub settlement_deadline: i64,
    pub timestamp: i64,
}

/// Event emitted when payout is claimed
#[event]
pub struct PayoutClaimed {
    pub market: Pubkey,
    pub claimer: Pubkey,
    pub position: Pubkey,
    pub amount_claimed: u64,
    pub shares_used: u64,
    pub timestamp: i64,
}

/// Event emitted when dispute is opened
#[event]
pub struct DisputeOpened {
    pub market: Pubkey,
    pub dispute_id: u64,
    pub disputer: Pubkey,
    pub original_outcome: u8,
    pub proposed_outcome: u8,
    pub stake_amount: u64,
    pub reason: String,
    pub timestamp: i64,
}

/// Event emitted when dispute is resolved
#[event]
pub struct DisputeResolved {
    pub market: Pubkey,
    pub dispute_id: u64,
    pub resolver: Pubkey,
    pub resolution: String,
    pub final_outcome: u8,
    pub timestamp: i64,
}

File: apps/indexer/src/types/events.ts
import { PublicKey } from '@solana/web3.js';

/**
 * Market Factory Events
 */

export interface MarketCreatedEvent {
  marketId: bigint;
  market: PublicKey;
  creator: PublicKey;
  questionText: string;
  numOutcomes: number;
  tradingFeeBps: number;
  resolutionDeadline: bigint;
  timestamp: bigint;
}

export interface LiquidityAddedEvent {
  market: PublicKey;
  provider: PublicKey;
  amount: bigint;
  lpTokensMinted: bigint;
  timestamp: bigint;
}

export interface LiquidityRemovedEvent {
  market: PublicKey;
  provider: PublicKey;
  lpTokensBurned: bigint;
  amountReceived: bigint;
  timestamp: bigint;
}

export interface TradeExecutedEvent {
  market: PublicKey;
  trader: PublicKey;
  outcomeIndex: number;
  isBuy: boolean;
  amount: bigint;
  price: bigint;
  fee: bigint;
  shares: bigint;
  timestamp: bigint;
}

export interface MarketStateChangedEvent {
  market: PublicKey;
  oldState: string;
  newState: string;
  authority: PublicKey;
  timestamp: bigint;
}

export interface MarketResolvedEvent {
  market: PublicKey;
  winningOutcome: number;
  resolver: PublicKey;
  timestamp: bigint;
}

export interface MarketClosedEvent {
  market: PublicKey;
  creator: PublicKey;
  feesWithdrawn: bigint;
  totalVolume: bigint;
  totalTrades: bigint;
  timestamp: bigint;
}

export interface FeesWithdrawnEvent {
  market: PublicKey;
  recipient: PublicKey;
  amount: bigint;
  timestamp: bigint;
}

/**
 * Market Settlement Events
 */

export interface ProofSubmittedEvent {
  market: PublicKey;
  proofId: bigint;
  submitter: PublicKey;
  outcomeIndex: number;
  proofData: string;
  timestamp: bigint;
}

export interface ProofVerifiedEvent {
  market: PublicKey;
  proofId: bigint;
  verifier: PublicKey;
  isValid: boolean;
  timestamp: bigint;
}

export interface SettlementMarketResolvedEvent {
  market: PublicKey;
  resolver: PublicKey;
  winningOutcome: number;
  totalPayout: bigint;
  settlementDeadline: bigint;
  timestamp: bigint;
}

export interface PayoutClaimedEvent {
  market: PublicKey;
  claimer: PublicKey;
  position: PublicKey;
  amountClaimed: bigint;
  sharesUsed: bigint;
  timestamp: bigint;
}

export interface DisputeOpenedEvent {
  market: PublicKey;
  disputeId: bigint;
  disputer: PublicKey;
  originalOutcome: number;
  proposedOutcome: number;
  stakeAmount: bigint;
  reason: string;
  timestamp: bigint;
}

export interface DisputeResolvedEvent {
  market: PublicKey;
  disputeId: bigint;
  resolver: PublicKey;
  resolution: string;
  finalOutcome: number;
  timestamp: bigint;
}

/**
 * Union type of all events
 */
export type SolanaEvent =
  | MarketCreatedEvent
  | LiquidityAddedEvent
  | LiquidityRemovedEvent
  | TradeExecutedEvent
  | MarketStateChangedEvent
  | MarketResolvedEvent
  | MarketClosedEvent
  | FeesWithdrawnEvent
  | ProofSubmittedEvent
  | ProofVerifiedEvent
  | SettlementMarketResolvedEvent
  | PayoutClaimedEvent
  | DisputeOpenedEvent
  | DisputeResolvedEvent;

/**
 * Event type enum for discriminated unions
 */
export enum EventType {
  MarketCreated = 'MarketCreated',
  LiquidityAdded = 'LiquidityAdded',
  LiquidityRemoved = 'LiquidityRemoved',
  TradeExecuted = 'TradeExecuted',
  MarketStateChanged = 'MarketStateChanged',
  MarketResolved = 'MarketResolved',
  MarketClosed = 'MarketClosed',
  FeesWithdrawn = 'FeesWithdrawn',
  ProofSubmitted = 'ProofSubmitted',
  ProofVerified = 'ProofVerified',
  SettlementMarketResolved = 'SettlementMarketResolved',
  PayoutClaimed = 'PayoutClaimed',
  DisputeOpened = 'DisputeOpened',
  DisputeResolved = 'DisputeResolved',
}

/**
 * Parsed event with metadata
 */
export interface ParsedEvent {
  type: EventType;
  data: SolanaEvent;
  signature: string;
  slot: number;
  blockTime: number | null;
}
